# --> Import NLTK library

In [ ]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\abhay\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\abhay\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [ ]:
import pandas as pd

In [ ]:
df= pd.read_csv("FakeNewsNet.csv")

In [ ]:
df

title  \
0      Kandi Burruss Explodes Over Rape Accusation on...   
1      People's Choice Awards 2018: The best red carp...   
2      Sophia Bush Sends Sweet Birthday Message to 'O...   
3      Colombian singer Maluma sparks rumours of inap...   
4      Gossip Girl 10 Years Later: How Upper East Sid...   
...                                                  ...   
23191  Pippa Middleton wedding: In case you missed it...   
23192  Zayn Malik & Gigi Hadid’s Shocking Split: Why ...   
23193  Jessica Chastain Recalls the Moment Her Mother...   
23194  Tristan Thompson Feels "Dumped" After Khloé Ka...   
23195  Kelly Clarkson Performs a Medley of Kendrick L...   

                                                news_url  \
0      http://toofab.com/2017/05/08/real-housewives-a...   
1      https://www.today.com/style/see-people-s-choic...   
2      https://www.etonline.com/news/220806_sophia_bu...   
3      https://www.dailymail.co.uk/news/article-33655...   
4      https://www.zerchoo.com/entertainment/gossip-g...   
...                                                  ...   
23191  https://www.express.co.uk/news/royal/807049/pi...   
23192  hollywoodlife.com/2018/03/13/zayn-malik-gigi-h...   
23193  http://www.justjared.com/2018/01/17/jessica-ch...   
23194  www.intouchweekly.com/posts/tristan-thompson-f...   
23195  https://www.billboard.com/articles/news/bbma/8...   

               source_domain  tweet_num  real  
0                 toofab.com         42     1  
1              www.today.com          0     1  
2           www.etonline.com         63     1  
3        www.dailymail.co.uk         20     1  
4            www.zerchoo.com         38     1  
...                      ...        ...   ...  
23191      www.express.co.uk         52     1  
23192      hollywoodlife.com          7     0  
23193      www.justjared.com         26     1  
23194  www.intouchweekly.com         24     0  
23195      www.billboard.com         85     1  

[23196 rows x 5 columns]

In [ ]:
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize

In [ ]:
import re    #rezx library name

In [ ]:
df.head()

title  \
0  Kandi Burruss Explodes Over Rape Accusation on...   
1  People's Choice Awards 2018: The best red carp...   
2  Sophia Bush Sends Sweet Birthday Message to 'O...   
3  Colombian singer Maluma sparks rumours of inap...   
4  Gossip Girl 10 Years Later: How Upper East Sid...   

                                            news_url        source_domain  \
0  http://toofab.com/2017/05/08/real-housewives-a...           toofab.com   
1  https://www.today.com/style/see-people-s-choic...        www.today.com   
2  https://www.etonline.com/news/220806_sophia_bu...     www.etonline.com   
3  https://www.dailymail.co.uk/news/article-33655...  www.dailymail.co.uk   
4  https://www.zerchoo.com/entertainment/gossip-g...      www.zerchoo.com   

   tweet_num  real  
0         42     1  
1          0     1  
2         63     1  
3         20     1  
4         38     1

# ---> Checks Null Values

In [ ]:
df.isnull().sum()

title              0
news_url         330
source_domain    330
tweet_num          0
real               0
dtype: int64

# --> Checks Duplicate Values

In [ ]:
df.duplicated().sum()

np.int64(137)

# --> Drop both columns

In [ ]:
df.drop(columns=['news_url','tweet_num'],axis=1,inplace=True)

In [ ]:
df.drop_duplicates(keep='first',inplace=True)

In [ ]:
stop_words = set(stopwords.words('english'))
len(stop_words)

198

In [ ]:
def remove_stopwords(text):
    words=word_tokenize(text)
    return ' '.join([word for word in words if word.lower() not in stop_words])

In [ ]:
df['title'] = df['title'].apply(remove_stopwords)

In [ ]:
import nltk
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\abhay\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [ ]:
def clean_title(line):
    line = str(line)
    line = re.sub(r'[^a-zA-Z\s]', '' ,line)
    line = line.lower()
    return line

In [ ]:
df['title'] = df['title'].apply(clean_title)

In [ ]:
df['source_domain'] = df['source_domain'].apply(clean_title)

In [ ]:
df['text'] = df['title'] + ' ' + df['source_domain']

In [ ]:
df.head()

title     source_domain  real  \
0  kandi burruss explodes rape accusation real ho...         toofabcom     1   
1     people s choice awards   best red carpet looks       wwwtodaycom     1   
2  sophia bush sends sweet birthday message one t...    wwwetonlinecom     1   
3  colombian singer maluma sparks rumours inappro...  wwwdailymailcouk     1   
4  gossip girl  years later  upper east siders sh...     wwwzerchoocom     1   

                                                text  
0  kandi burruss explodes rape accusation real ho...  
1  people s choice awards   best red carpet looks...  
2  sophia bush sends sweet birthday message one t...  
3  colombian singer maluma sparks rumours inappro...  
4  gossip girl  years later  upper east siders sh...

In [ ]:
x= df['text'].values
y = df['real'].values

# --> Using Train_Test_Split

In [ ]:
from sklearn.model_selection import train_test_split

# --> Using TfidVectorizer

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
import sklearn

In [ ]:
x_train , x_test ,y_train , y_test = train_test_split(x,y ,test_size=1/3,random_state=42)

In [ ]:
vectorizer = TfidfVectorizer()

In [ ]:
x_train = vectorizer.fit_transform(x_train)
x_test = vectorizer.transform(x_test)

# --> Using LogisticRegression

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
logistic = LogisticRegression()

In [ ]:
logistic.fit(x_train,y_train)

LogisticRegression()

In [ ]:
pred = logistic.predict(x_test)

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy_score(y_test,pred)

0.8593364409103372

# --> Using DecisionTree

In [ ]:
from sklearn import tree

In [ ]:
tree1 = tree.DecisionTreeClassifier()

In [ ]:
tree1.fit(x_train,y_train)

DecisionTreeClassifier()

In [ ]:
pred1 = tree1.predict(x_test)

In [ ]:
accuracy_score(pred1,y_test)

0.7931176309295311

# --> Using RandomForestClassifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
Random = RandomForestClassifier()

In [ ]:
Random.fit(x_train,y_train)

RandomForestClassifier()

In [ ]:
pred2 = Random.predict(x_test)

In [ ]:
accuracy_score(pred2,y_test)

0.8524814916369619

# --> Using XGBoost

In [ ]:
import xgboost

In [ ]:
xgb = xgboost.XGBClassifier()

In [ ]:
xgb.fit(x_train,y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              feature_weights=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=None, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=None,
              n_jobs=None, num_parallel_tree=None, ...)

In [ ]:
pred3 = xgb.predict(x_test)

In [ ]:
accuracy_score(pred3,y_test)

0.8539895804771045

# --> We using 4 models to check accuracy_score in this data
    # LogisticRegression = 0.85
    # DecisionTree = 0.79
    # RandomforestClassifier = 0.85
    # XGboost Classifier = 0.85

# --> Best accuracy score is LogisticRegression model score = 0.85